pkg/reco/reco.go

```diff
@@ -0,0 +1,237 @@
+package reco
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	rolloutv1alpha1 "github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/flipkart-incubator/ottoscalr/api/v1alpha1"
+	"github.com/flipkart-incubator/ottoscalr/pkg/metrics"
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"math"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type Recommender interface {
+	Recommend(workloadSpec v1alpha1.WorkloadSpec) (*v1alpha1.HPAConfiguration, error)
+}
+
+type CpuUtilizationBasedRecommender struct {
+	k8sClient    client.Client
+	redLineUtil  float64
+	metricWindow time.Duration
+	scraper      metrics.Scraper
+	metricStep   time.Duration
+	minTarget    int
+	maxTarget    int
+	logger       logr.Logger
+}
+
+func NewCpuUtilizationBasedRecommender(k8sClient client.Client,
+	redLineUtil float64,
+	metricWindow time.Duration,
+	scraper metrics.Scraper,
+	metricStep time.Duration,
+	minTarget int,
+	maxTarget int,
+	logger logr.Logger) *CpuUtilizationBasedRecommender {
+	return &CpuUtilizationBasedRecommender{
+		k8sClient:    k8sClient,
+		redLineUtil:  redLineUtil,
+		metricWindow: metricWindow,
+		scraper:      scraper,
+		metricStep:   metricStep,
+		minTarget:    minTarget,
+		maxTarget:    maxTarget,
+		logger:       logger,
+	}
+}
+
+func (c *CpuUtilizationBasedRecommender) Recommend(workloadSpec v1alpha1.WorkloadSpec) (*v1alpha1.HPAConfiguration,
+	error) {
+
+	end := time.Now()
+	start := end.Add(c.metricWindow)
+
+	dataPoints, err := c.scraper.GetAverageCPUUtilizationByWorkload(workloadSpec.Namespace,
+		workloadSpec.Name,
+		start,
+		end,
+		c.metricStep)
+	if err != nil {
+		c.logger.Error(err, "Error while scraping GetCPUUtilizationBreachDataPoints.")
+		return nil, nil
+	}
+
+	acl, err := c.scraper.GetACLByWorkload(workloadSpec.Namespace, workloadSpec.Name)
+	if err != nil {
+		c.logger.Error(err, "Error while getting GetACL.")
+		return nil, nil
+	}
+
+	perPodResources, err := c.getContainerCPULimitsSum(workloadSpec.Namespace, workloadSpec.Kind, workloadSpec.Name)
+	if err != nil {
+		c.logger.Error(err, "Error while getting getContainerCPULimitsSum")
+		return nil, err
+	}
+
+	optimalTargetUtil, minReplicas, maxReplicas, err := c.findOptimalTargetUtilization(dataPoints,
+		acl,
+		c.minTarget,
+		c.maxTarget,
+		perPodResources)
+	if err != nil {
+		c.logger.Error(err, "Error while executing findOptimalTargetUtilization")
+		return nil, err
+	}
+
+	return &v1alpha1.HPAConfiguration{Min: minReplicas, Max: maxReplicas, TargetMetricValue: optimalTargetUtil}, nil
+}
+
+type TimerEvent struct {
+	Timestamp time.Time
+	Delta     float64
+}
+
+func (c *CpuUtilizationBasedRecommender) simulateHPA(dataPoints []metrics.DataPoint,
```

Comment:
Please add some comments to this method to elaborate on the approach/steps.

Reply:
Done.